core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -30,7 +30,11 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
 
   def shutdown(): Unit = {
     initiateShutdown()
-    awaitShutdown()
+    if (this.isAlive) {
+      awaitShutdown()
+    } else {
+      shutdownComplete.countDown()
```

Comment:
In terms of usage, a user could call (1) shutdown(), or (2) initiateShutdown(); some other code; awaitShutdown(). In the case when a thread is never started, the patch makes sure that (1) doesn't block. For consistency, it seems that we should make (2) not blocking too?

Reply:
Hmm I'm not sure, my interpretation of `awaitShutdown()` was that it should always block the caller until shutdown is complete, regardless of the thread being started or not.

Looking through the codebase it seems like it's mostly used when we know the thread is already started, so it's probably best to do the `this.isAlive` check in `awaitShutdown()` in order to make both (1) and (2) non-blocking.